<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ amlpmpsupport</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">amlpmpsupport</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/amlpmpsupport-examples.html">amlpmpsupport - Examples</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="add_entrez_ids_from_hugo_genes.html">add_entrez_ids_from_hugo_genes()</a></code> </p>
        </td>
        <td><p>Add Entrez gene IDs to a data frame of voom hits</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="apply_eln2015_cyto.html">apply_eln2015_cyto()</a></code> </p>
        </td>
        <td><p>Apply ELN2015 stratification based on cytogenetics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="apply_eln2017_cyto.html">apply_eln2017_cyto()</a></code> </p>
        </td>
        <td><p>Apply ELN2017 stratification based on cytogenetics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="apply_eln2017_rna.html">apply_eln2017_rna()</a></code> </p>
        </td>
        <td><p>Apply ELN2017-RNA stratification</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="apply_expression_signature_reclassification.html">apply_expression_signature_reclassification()</a></code> </p>
        </td>
        <td><p>Update an existing stratification by applying a gene expression score</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="aps_volcano_plot.html">aps_volcano_plot()</a></code> </p>
        </td>
        <td><p>Draw Volcano Plot for APS DE Comparison</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calculate_reference_set_coefficient_of_variation.html">calculate_reference_set_coefficient_of_variation()</a></code> </p>
        </td>
        <td><p>Calculate reference set coefficient of variation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calculate_reference_set_mean_expression.html">calculate_reference_set_mean_expression()</a></code> </p>
        </td>
        <td><p>Calculate reference set mean expression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="connect_to_aml_project_orm.html">connect_to_aml_project_orm()</a></code> </p>
        </td>
        <td><p>Connect to AML Project ORM Database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="construct_rna_qc_dataframe.html">construct_rna_qc_dataframe()</a></code> </p>
        </td>
        <td><p>Construct a data frame of RNA-SeQC results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="construct_sailfish_gene_dataframe.html">construct_sailfish_gene_dataframe()</a></code> </p>
        </td>
        <td><p>Construct a data frame of Sailfish gene-level results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convert_sailfish_df_to_matrix.html">convert_sailfish_df_to_matrix()</a></code> </p>
        </td>
        <td><p>From a data frame of Sailfish gene-level results, construct an expression matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="density_plotter.html">density_plotter()</a></code> </p>
        </td>
        <td><p>Generate a density-plot of expression of selected genes, facetted by selected feature</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="downregulated_genes.html">downregulated_genes()</a></code> </p>
        </td>
        <td><p>downregulated_genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="eng_log_breaks.html">eng_log_breaks()</a></code> </p>
        </td>
        <td><p>Generate a set of breaks using the 'Engineer's log scale'</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="expression_matrix_to_long_df.html">expression_matrix_to_long_df()</a></code> </p>
        </td>
        <td><p>Transform expression matrix to long dataframe</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fit_coxph.html">fit_coxph()</a></code> </p>
        </td>
        <td><p>Fit a Cox proportional hazards model</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gather_expression_matrix_to_tidy_df.html">gather_expression_matrix_to_tidy_df()</a></code> </p>
        </td>
        <td><p>Convert an expression matrix to a tidy DF</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_cluster_assignments_for_k.html">get_cluster_assignments_for_k()</a></code> </p>
        </td>
        <td><p>Retrieve cluster assignments for a given value of k</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ggsave_pdf_and_png.html">ggsave_pdf_and_png()</a></code> </p>
        </td>
        <td><p>Save plot to PDF and PNG</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="grid_arrange_shared_legend.html">grid_arrange_shared_legend()</a></code> </p>
        </td>
        <td><p>Generate a grid of plots with a shared legend</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="group_tally_table.html">group_tally_table()</a></code> </p>
        </td>
        <td><p>Group / Tally / Table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="hit_plotter.html">hit_plotter()</a></code> </p>
        </td>
        <td><p>Generate a dot-plot of expression of selected genes, facetted by selected feature</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="limma_logFC_to_signed_foldchange.html">limma_logFC_to_signed_foldchange()</a></code> </p>
        </td>
        <td><p>Convert limma's logFC value to signed fold-change</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="log2_plus1_transform.html">log2_plus1_transform()</a></code> </p>
        </td>
        <td><p>Transform data by taking the log2(TPM+1)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="long_df_to_ranked_matrix.html">long_df_to_ranked_matrix()</a></code> </p>
        </td>
        <td><p>Transform long dataframe to sorted matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="matrix_glimpse.html">matrix_glimpse()</a></code> </p>
        </td>
        <td><p># Report basic information about a matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="munge_voom_to_ipa.html">munge_voom_to_ipa()</a></code> </p>
        </td>
        <td><p>Munge differential expression hits in voom format to IPA-friendly format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="nest_libraries_by_patient.html">nest_libraries_by_patient()</a></code> </p>
        </td>
        <td><p>Add nested library information to a patient-level data frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="num_to_text.html">num_to_text()</a></code> </p>
        </td>
        <td><p>Convert a small number to a text representation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pathway_enrichment_plot.html">pathway_enrichment_plot()</a></code> </p>
        </td>
        <td><p>Generate a plot of enriched pathways</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_heatmap_for_k.html">plot_heatmap_for_k()</a></code> </p>
        </td>
        <td><p>Plot a pheatmap for a single value of k</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_rf_predictions.html">plot_rf_predictions()</a></code> </p>
        </td>
        <td><p>Plot predictions from a randomForest model</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_survival.html">plot_survival()</a></code> </p>
        </td>
        <td><p>Plot survival curves</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_vars_by_coverage.html">plot_vars_by_coverage()</a></code> </p>
        </td>
        <td><p>This function takes the long and wide data frames, and plots all variants by covearge, colouring them by concordance status</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_vars_by_vaf.html">plot_vars_by_vaf()</a></code> </p>
        </td>
        <td><p>Join to the plotting frame and re-plot for VAF</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pretty_kable.html">pretty_kable()</a></code> </p>
        </td>
        <td><p>Set a function for pretty-printing kables</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pretty_num.html">pretty_num()</a></code> </p>
        </td>
        <td><p>Set a function for pretty-printing numbers in the text</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pull_enrichr_enriched_terms.html">pull_enrichr_enriched_terms()</a></code> </p>
        </td>
        <td><p>Pull EnrichR enriched terms for a single pathway source</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pull_enrichr_terms_to_plot.html">pull_enrichr_terms_to_plot()</a></code> </p>
        </td>
        <td><p>Identify enriched terms across cohorts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="random_forest_summary.html">random_forest_summary()</a></code> </p>
        </td>
        <td><p>From a random forest model generated by caret, generate summary plots</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_filled_variant_tsv.html">read_filled_variant_tsv()</a></code> </p>
        </td>
        <td><p>This function reads in the 'long' TSV variant sheets generated by the makefile</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rescale_genes_by_reference.html">rescale_genes_by_reference()</a></code> </p>
        </td>
        <td><p>Rescale all genes by reference expression levels</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="retrieve_db_comments.html">retrieve_db_comments()</a></code> </p>
        </td>
        <td><p>Retrieve DB Comments</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="retrieve_db_curated_results.html">retrieve_db_curated_results()</a></code> </p>
        </td>
        <td><p>Retrieve DB Curated Results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="retrieve_db_libraries.html">retrieve_db_libraries()</a></code> </p>
        </td>
        <td><p>Retrieve DB Libraries</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="retrieve_db_paths.html">retrieve_db_paths()</a></code> </p>
        </td>
        <td><p>Retrieve DB Paths</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="retrieve_db_specimens.html">retrieve_db_specimens()</a></code> </p>
        </td>
        <td><p>Retrieve DB Specimens</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rf_predictions.html">rf_predictions()</a></code> </p>
        </td>
        <td><p>Extract predictions from a random forest model</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_cna_workflow.html">run_cna_workflow()</a></code> </p>
        </td>
        <td><p>Run the CNA workflow</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_gage_for_feature.html">run_gage_for_feature()</a></code> </p>
        </td>
        <td><p>Run GAGE gene-set enrichment analysis for a given feature</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_voom_for_feature.html">run_voom_for_feature()</a></code> </p>
        </td>
        <td><p>Run voom for a given feature comparison</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scale_by_rolling_mean.html">scale_by_rolling_mean()</a></code> </p>
        </td>
        <td><p>Scale expression estimates by rolling mean</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="select_by_position.html">select_by_position()</a></code> </p>
        </td>
        <td><p>Quick function for mapping to select out list items</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="split_gatk_format_vals.html">split_gatk_format_vals()</a></code> </p>
        </td>
        <td><p>Split a FORMAT string from GATK-haplotype to get simpler statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spread_expression_df_to_matrix.html">spread_expression_df_to_matrix()</a></code> </p>
        </td>
        <td><p>Convert a 'tidy' expression DF to a matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="spread_filled_snv_df_to_wide.html">spread_filled_snv_df_to_wide()</a></code> </p>
        </td>
        <td><p>This function takes the 'long, filled' data frames with variant observations, and converts them to 'wide' format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subset_genes_by_thresholds.html">subset_genes_by_thresholds()</a></code> </p>
        </td>
        <td><p>Subset expression matrix by thresholds</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subset_genes_for_plots.html">subset_genes_for_plots()</a></code> </p>
        </td>
        <td><p>Subset the TPM matrix to the genes of interest</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subset_heatmap.html">subset_heatmap()</a></code> </p>
        </td>
        <td><p>Generate a heatmap for a subset of genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subset_pc_plot.html">subset_pc_plot()</a></code> </p>
        </td>
        <td><p>Generate a prinicple component plot for a subset of genes, coloured by a given feature</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subtype_to_disease.html">subtype_to_disease()</a></code> </p>
        </td>
        <td><p>Convert subtype indications to disease status</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tfl_split_sort.html">tfl_split_sort()</a></code> </p>
        </td>
        <td><p>A function to sort TFL IDs into ascending order</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="train_caret_rf.html">train_caret_rf()</a></code> </p>
        </td>
        <td><p>Train a random forest model with caret</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="upregulated_genes.html">upregulated_genes()</a></code> </p>
        </td>
        <td><p>upregulated_genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="volcano_plotter.html">volcano_plotter()</a></code> </p>
        </td>
        <td><p>Generate a volcano plot of expression hits</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Rod Docking.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


