#' This function reads in the 'long' TSV variant sheets generated by the makefile
#'
#' @param tsv_file
#'
#' @return filled.df A data frame containing filled SNV data
#' @export
#'
#' @examples
read_filled_variant_tsv <- function(tsv_file) {
  filled.df <- readr::read_tsv(tsv_file,
                        col_names = c('chr', 'pos', 'ref', 'alt', 'sample', 'gene',
                                      'transcript', 'protein', 'genotype',
                                      'hq_depth', 'vaf')) %>%
    dplyr::mutate(vaf_numeric = as.numeric(gsub('%', '', vaf)),
           var_key = paste(chr, pos, ref, alt, sep='_'),
           var_hgvs = paste(gene, transcript, protein, sep=';'),
           bool_genotype = as.integer(if_else(genotype %in% c('0/1', '1/1', '1|1', '0|1'),
                                              1, 0)))
  return(filled.df)
}

#' This function takes the 'long, filled' data frames with variant observations, and converts them to 'wide' format
#'
#' @param df
#'
#' @return
#' @export
#'
#' @examples
spread_filled_snv_df_to_wide <- function(df) {
  df %>%
    dplyr::mutate(concordance_col = 'black') %>%
    dplyr::select(-genotype, -hq_depth, -vaf, -vaf_numeric) %>%
    tidyr::spread(sample, bool_genotype) %>%
    as.data.frame() %>%
    return()
}

#' This function takes the long and wide data frames, and plots all variants by covearge, colouring them by concordance status
#'
#' @param wide.df
#' @param long.df
#' @param sample_name
#'
#' @return
#' @export
#'
#' @examples
plot_vars_by_coverage <- function(wide.df, long.df, sample_name) {
  title <- paste0("Coverage Depth for ", sample_name, " Replicate SNVs")
  wide.df %>%
    add_colour_vector_to_concordance() %>%
    dplyr::select(var_key, concordance_col) %>%
    dplyr::left_join(long.df, by = "var_key") %>%
    ggplot(aes(x = factor(var_key),
               y = hq_depth,
               shape = genotype,
               colour = concordance_col)) +
    geom_point(size=3) + scale_colour_identity() +
    scale_y_log10(breaks = c(1,2,5,10,20,50,100,200,500,1000, 2000)) +
    theme(axis.text.x=element_text(angle = 90, hjust = 0)) +
    coord_flip() +
    labs(title = title,
         x = "Variant",
         y = "High-quality Depth") %>%
    return()
}


#' Join to the plotting frame and re-plot for VAF
#'
#' @param wide.df
#' @param long.df
#' @param sample_name
#'
#' @return
#' @export
#'
#' @examples
plot_vars_by_vaf <- function(wide.df, long.df, sample_name) {
  title <- paste0("Variant allele fraction for ", sample_name, " Replicate SNVs")
  wide.df %>%
    add_colour_vector_to_concordance() %>%
    dplyr::select(var_key, concordance_col) %>%
    dplyr::left_join(long.df, by = "var_key") %>%
    ggplot(aes(x = factor(var_key),
               y = vaf_numeric,
               shape = genotype,
               colour = concordance_col)) +
    geom_point(size=3) + scale_colour_identity() +
    theme(axis.text.x=element_text(angle = 90, hjust = 0)) +
    coord_flip() +
    labs(title = title,
         x = "Variant",
         y = "Variant Allele Fraction") %>%
    return()
}
